---
title: "Process template data"
output: html_notebook
---
## Packages
```{r, warning=F}
library(tidyverse)
library(parallel)
library(stringdist)
```

# Create directory structures and download all sequencing data in 1-original folder
```{bash}
mkdir -p ../1-original ../2-trim ../3-deduplicate ../4-count ../5-final
```

# Process template data
### Trim template data with cutadapt 1.16
```{bash}
for file in ../1-original/template*.gz; do
  name=`echo ${file##*/}`
  name_sans_ext=`echo ${file##*/} | sed 's/.fastq.gz//'`

  echo ----------$name---------- >> ../2-trim/cutadapt_report.txt
  cutadapt -e 0.3 -j 40 -M 85 --match-read-wildcards -a TGGAATTCTCGGGTGCCAACTCCAGTCAC --discard-untrimmed -o ../2-trim/$name_sans_ext.trim.fq $file &>> ../2-trim/cutadapt_report.txt
done
```

### Deduplicate reads using dedupe.sh from BBtools
```{bash}
for file in ../2-trim/template*.fq; do
  name=`echo ${file##*/}`
  deduped=`echo ${file##*/} | sed 's/.trim.fq/.dd.fq/'`
  dedupedfile="../3-deduplicate/"$deduped

  echo ----------$name---------- >>../3-deduplicate/dedupe_report.txt
  /home/prshah/Software/pckgs/bbmap/dedupe.sh in=$file out=$dedupedfile s=2 nam=3 t=40 -Xmx100g &>>../3-deduplicate/dedupe_report.txt
done
```

### Trim template data with cutadapt 1.16
```{bash}
for file in ../3-deduplicate/template*.fq; do
  name=`echo ${file##*/}`
  deduped=`echo ${file##*/} | sed 's/.dd.fq/.dd.trim.fq/'`

  echo ----------$name---------- >> ../3-deduplicate/cutadapt_report.txt
  cutadapt -e 0.3 -j 40 -M 48 -g GCTCGTATAATGTGTGGAA --discard-untrimmed -o ../3-deduplicate/$deduped $file &>> ../3-deduplicate/cutadapt_report.txt
done
```

### Extract left and right barcodes from template data
```{bash}
for file in ../3-deduplicate/template*.trim.fq; do
  name=`echo ${file##*/}`
  barcode=`echo ${file##*/} | sed 's/.dd.trim.fq/.bc.map.tsv/'`
  unq="../4-count/"$barcode

#  python3 itc.template.finalbc.py $file $unq 

  unq2=`echo $unq | sed 's/.bc.map.tsv/.count.bc.map.tsv/' `
  tail -n +2 $unq | cut -f2,5 | sort | uniq -c | awk '{print $(NF-2)" "$(NF-1)" "$NF}' | sort -k2 > $unq2
done
```


```{r}
# Get template data filenames
templ.list <- list.files("../4-count/", "template.counts.*\\.count.bc.map.tsv", ignore.case = T, full.names = T)
rep.ids <- strsplit(templ.list, "counts.|.ITC") %>% sapply(., `[[`, 2)

# Read in template data
templ.data <- lapply(templ.list, read_delim, delim = " ", col_names = F)
templ.data <- lapply(templ.data, function(x) x %>% 
                       dplyr::rename(count = X1, n11 = X2, n14 = X3))

# Get number of unique N11 for each N14
n14.cnt.n11 <- function(temp.data){
  temp.data %>%
  dplyr::select(-count) %>% 
  group_by(n14) %>% 
  dplyr::summarise(n = n_distinct(n11))
}

n14.cnt.11.list <- lapply(templ.data, n14.cnt.n11)
  
unq.n11.n14 <- lapply(1:3, function(x) left_join(n14.cnt.11.list[[x]] %>% 
                                                   filter(n == 1),
                                                 templ.data[[x]], by = "n14"))
non.unq.n11.n14 <- lapply(1:3, function(x) left_join(n14.cnt.11.list[[x]] %>% 
                                                       filter(n > 1),
                                                 templ.data[[x]], by = "n14"))

# For n14 with multiple n11, fix any sequencing errors
fixn11 <- function(qc, q11, q14){
  tmp1 <- data.frame(q11, q14, qc, stringsAsFactors = F)
  tmp2 <- tmp1 %>% 
    group_by(q11) %>% 
    dplyr::summarise(weight = sum(qc))
  
  d <- c(stringdistmatrix(tmp2$q11))
  if(max(d)<=4 | (max(tmp2$weight) >= 10)){
    true.seq <- tmp2$q11[which.max(tmp2$weight)]
  }else{
    true.seq <- "NA"
  }
}

fix.non.unq.n11.n14 <- mclapply(non.unq.n11.n14, function(x){
    x %>% 
    group_by(n14) %>%
    mutate(newn11 = fixn11(q14 = n14, 
                           q11 = n11,
                           qc = count)) %>%
    ungroup() %>%
    mutate(n11 = newn11) %>% 
    dplyr::select(-newn11) %>%
    filter(n11 != "NA") %>%
    dplyr::select(-n) %>%
    group_by(n14, n11) %>%
    dplyr::summarise(count = sum(count)) %>%
    ungroup()
}, mc.cores = 3)

lapply(1:3, function(x){write_delim(bind_rows(unq.n11.n14[[x]] %>%
                                               dplyr::select(-n) %>%
                                               dplyr::select(n14, n11, count),
                                             fix.non.unq.n11.n14[[x]]) %>%
                                      dplyr::rename(N14 = n14, N11 = n11), paste0("../5-final/template.rep",x,".counts.tsv"), delim = "\t")})
```
